lib/ua_update_messaging.py

```diff
@@ -0,0 +1,167 @@
+present updated text about Ubuntu Advantage service and token state.
+"""
+
+import enum
+import logging
+import os
+
+try:
+    from typing import Optional, Tuple  # noqa
+except ImportError:
+    # typing isn't available on trusty, so ignore its absence
+    pass
+
+from uaclient.cli import setup_logging
+from uaclient import config
+from uaclient import entitlements
+from uaclient import defaults
+from uaclient import status
+from uaclient import util
+
+
+# Type of message file used for external messaging (APT and MOTD)
+@enum.unique
+class ExternalMessage(enum.Enum):
+    CONTRACT_EXPIRY_MOTD = "contract-expiry-status.tmpl"
+    CONTRACT_EXPIRY_APT = "contract-expiry-status-apt.tmpl"
+    ESM_ANNOUNCE = "esm-announce"
+
+
+def format_contract_expiry_messages(cfg: config.UAConfig) -> "Tuple[str, str]":
+    """Return warning message about contract expiry used by MOTD and APT.
+
+    Compare UAConfig.contract_remainining_days versus pending expiry period,
+    expiry grace period and absolute expiry dates. Construct appropriate
+    message to emit in APT and MOTD based on expired contract state.
+
+    :param cfg: UAConfig instance for this environment.
+
+    :return: Tuple of strings for contract expiry tailored for MOTD and APT.
+        Empty string for items in the tuple when no message needed.
+    """
+    motd_msg = ""
+    apt_msg = ""
+    if not cfg.is_attached:
+        return motd_msg, apt_msg
+    grace_period = defaults.CONTRACT_EXPIRY_GRACE_PERIOD_DAYS
+    pending_expiry = defaults.CONTRACT_EXPIRY_PENDING_DAYS
+    remaining_days = cfg.contract_remaining_days
+    if 0 <= remaining_days <= pending_expiry:
+        motd_msg = status.MESSAGE_CONTRACT_NEAR_EXPIRED_TMPL.format(
+            remaining_days=remaining_days, url=defaults.BASE_UA_URL
+        )
+        apt_msg = motd_msg
+    elif -grace_period <= remaining_days < 0:
+        motd_msg = status.MESSAGE_CONTRACT_EXPIRED_GRACE_PERIOD_TMPL.format(
+            expired_date=cfg.contract_expiry_datetime.strftime("%d %b %Y"),
+            remaining_days=remaining_days + grace_period,
+            url=defaults.BASE_UA_URL,
+        )
+        apt_msg = motd_msg
+    elif remaining_days < -grace_period:
+        if util.get_platform_info()["series"] in ("trusty", "xenial"):
+            no_warranty = status.MESSAGE_UBUNTU_NO_WARRANTY
+        else:
+            no_warranty = ""
+        motd_msg = (
+            status.MESSAGE_CONTRACT_EXPIRED_MOTD_TMPL.format(
+                num_pkg="{ESM_APPS_PKG_COUNT}", url=defaults.BASE_UA_URL
+            )
+            + no_warranty
+        )
+        apt_msg = (
+            status.MESSAGE_CONTRACT_EXPIRED_APT_TMPL.format(
+                num_pkg="{ESM_APPS_PKG_COUNT}", url=defaults.BASE_UA_URL
+            )
+            + no_warranty
+        )
+    return motd_msg, apt_msg
+
+
+def format_esm_announcement_message(cfg: config.UAConfig) -> "Optional[str]":
```

Comment:
@blackboxsw Another check we need to add here. If esm-apps is already enabled, we don't want to show the announcement message, right?

Reply:
Excellent point. will drop that message on enabled +1